dlt/common/destination/reference.py

```diff
@@ -342,59 +343,72 @@ def _verify_schema(self) -> None:
         * Removes and warns on (unbound) incomplete columns
         """
 
-        for table in self.schema.data_tables():
-            table_name = table["name"]
-            if len(table_name) > self.capabilities.max_identifier_length:
-                raise IdentifierTooLongException(
-                    self.config.destination_type,
-                    "table",
-                    table_name,
-                    self.capabilities.max_identifier_length,
-                )
+        def verify_merge_disposition(table: TTableSchema) -> None:
             if table.get("write_disposition") == "merge":
                 if "x-merge-strategy" in table and table["x-merge-strategy"] not in MERGE_STRATEGIES:  # type: ignore[typeddict-item]
                     raise SchemaException(
-                        f'"{table["x-merge-strategy"]}" is not a valid merge strategy. '  # type: ignore[typeddict-item]
+                        f'`{table["x-merge-strategy"]}` is not a valid merge strategy. '  # type: ignore[typeddict-item]
                         f"""Allowed values: {', '.join(['"' + s + '"' for s in MERGE_STRATEGIES])}."""
                     )
-                if (
-                    table.get("x-merge-strategy") == "delete-insert"
-                    and not has_column_with_prop(table, "primary_key")
-                    and not has_column_with_prop(table, "merge_key")
-                ):
-                    logger.warning(
-                        f"Table {table_name} has `write_disposition` set to `merge`"
-                        " and `merge_strategy` set to `delete-insert`, but no primary or"
-                        " merge keys defined."
-                        " dlt will fall back to `append` for this table."
-                    )
+                if table.get("x-merge-strategy") == "delete-insert":
+                    if not has_column_with_prop(table, "primary_key") and not has_column_with_prop(
+                        table, "merge_key"
+                    ):
+                        logger.warning(
+                            f"Table {table['name']} has `write_disposition` set to `merge`"
+                            " and `merge_strategy` set to `delete-insert`, but no primary or"
+                            " merge keys defined."
+                            " dlt will fall back to `append` for this table."
+                        )
+                elif table.get("x-merge-strategy") == "upsert":
+                    if self.config.destination_name not in ("postgres", "snowflake"):
```

Comment:
Should we add a `supported_merge_strategies` destination capability?

Reply:
yes - if you at some point (as early as possible ie. when passing data to normalizer when we must have destination capabilities) are able to issue a warning and say which strategy will be used instead.

same thing with replace strategies (we have 3 afaik)

also if you want to do it: we do not even store which write dispositions are supported. so maybe this is a separate ticket :)

Reply:
- added `supported_merge_strategies` capability and configured it for all SQL destinations
- added `_verify_destination_capabilities` that checks configured merge strategy against supported merge strategies
  - raises error if not supported
    - (I know you suggested falling back to a supported strategy and issueing a warning instead, but I'm not a fan of that approach. Will of course change it if that's how we do things in `dlt`, but wanted to challenge it first.)
  - called right before `normalize` step — is that the right place?

I can create a separate ticket for supported replace strategies and supported write dispositions capabilities.